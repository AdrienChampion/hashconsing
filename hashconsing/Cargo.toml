[package]
name = "hashconsing"
version = "1.6.0"
authors = ["Adrien Champion <adrien.champion@email.com>"]
description = "A hash consing library."
documentation = "https://docs.rs/hashconsing"
homepage = "https://github.com/AdrienChampion/hashconsing"
repository = "https://github.com/AdrienChampion/hashconsing"
readme = "README.md"
categories = [
    "caching",
    "compression",
    "concurrency",
    "data-structures",
    "memory-management",
]
keywords = ["hashconsing", "hash", "consing", "sharing", "caching"]
license = "MIT/Apache-2.0"
edition = "2021"

[package.metadata.docs.rs]
features = ["unstable_docrs"]

[features]
with_ahash = ["ahash"]
unstable_docrs = ["with_ahash"]
derive = ["hashconsing-derive"]

[dependencies]
lazy_static = "1.*"
hashconsing-derive = { version = "0.1.0", optional = true, path = "../hashconsing_derive" }

[dependencies.ahash]
version = "^0.8.3"
optional = true

[dev-dependencies]
crossbeam-utils = "=0.8.15"
trybuild = "^1.0"
rayon = "^1.5"
rand = "0.8"