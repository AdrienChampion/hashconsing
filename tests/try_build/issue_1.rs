//! From https://github.com/AdrienChampion/hashconsing/issues/1.

#![forbid(unsafe_code)]

use hashconsing::{HConsign, HashConsign};

use crossbeam_utils::thread;
use std::cell::Cell;
use std::hash::{Hash, Hasher};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
enum RefOrInt<'a> {
    Ref(&'a u64),
    Int(u64),
}

#[derive(PartialEq, Eq)]
struct HashableCell<T: Eq + PartialEq + Copy> {
    cell: Cell<T>,
}
// Fake hashing function just so we can get a HConsed going.
impl<T: Eq + PartialEq + Copy> Hash for HashableCell<T> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        1024.hash(state);
    }
}

static SOME_INT: u64 = 123;

fn main() {
    let cell = Cell::new(RefOrInt::Ref(&SOME_INT));
    let hashable_cell = HashableCell { cell: cell };

    let mut factory: HConsign<_> = HConsign::empty();
    let hcons_cell_ref = factory.mk(&hashable_cell);
    thread::scope(|s| {
        s.spawn(move |_| {
            let smuggled_cell = &hcons_cell_ref.get().cell;

            loop {
                // Repeatedly write Ref(&addr) and Int(0xdeadbeef) into the cell.
                smuggled_cell.set(RefOrInt::Ref(&SOME_INT));
                smuggled_cell.set(RefOrInt::Int(0xdeadbeef));
            }
        });

        loop {
            if let RefOrInt::Ref(addr) = hashable_cell.cell.get() {
                // Hope that between the time we pattern match the object as a
                // `Ref`, it gets written to by the other thread.
                if addr as *const u64 == &SOME_INT as *const u64 {
                    continue;
                }

                // Due to the data race, obtaining Ref(0xdeadbeef) is possible
                println!("Pointer is now: {:p}", addr);
                println!("Dereferencing addr will now segfault: {}", *addr);
            }
        }
    });
}
