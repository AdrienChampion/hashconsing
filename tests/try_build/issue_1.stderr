warning: unused import: `HConsed`
 --> $DIR/issue_1.rs:5:19
  |
5 | use hashconsing::{HConsed, HConsign, HashConsign};
  |                   ^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0277]: `std::cell::Cell<RefOrInt<'_>>` cannot be shared between threads safely
  --> $DIR/issue_1.rs:37:11
   |
37 |         s.spawn(move |_| {
   |           ^^^^^ `std::cell::Cell<RefOrInt<'_>>` cannot be shared between threads safely
   |
   = help: within `HashableCell<RefOrInt<'_>>`, the trait `std::marker::Sync` is not implemented for `std::cell::Cell<RefOrInt<'_>>`
   = note: required because it appears within the type `HashableCell<RefOrInt<'_>>`
   = note: required because of the requirements on the impl of `std::marker::Send` for `&HashableCell<RefOrInt<'_>>`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Arc<&HashableCell<RefOrInt<'_>>>`
   = note: required because it appears within the type `hashconsing::HConsed<&HashableCell<RefOrInt<'_>>>`
   = note: required because it appears within the type `[closure@$DIR/tests/try_build/issue_1.rs:37:17: 45:10 hcons_cell_ref:hashconsing::HConsed<&HashableCell<RefOrInt<'_>>>]`
